style: Sync changes from mozilla-central.

emilio opened the pull request:
See each individual commit for details.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/23325)
<!-- Reviewable:end -->

jfkthame committed:
style: Allow full-width and/or full-size-kana values of text-transform to be combined with a case transformation.

Differential Revision: https://phabricator.services.mozilla.com/D27402

emilio committed:
style: The counters code should use atoms rather than strings.

Servo already atomizes the counter names, it makes no sense to copy the string
rather than bumping the refcount.

Differential Revision: https://phabricator.services.mozilla.com/D27061

emilio committed:
style: Refactor the selector parser to make implementing ::part() easier.

::slotted() is already weird in the sense that it supports a pseudo-element
afterwards (so ::slotted(*)::before is valid for example).

::part() is weirder because you are supposed to allow stuff like
::part(foo):hover, ::part(foo):hover::before, etc.

In order to avoid making the already-complex parse_compound_selector more
complex, shuffle stuff so that we pass the progress of our current compound
selector around, and is the parsing code for each selector which decides whether
it's ok to parse at the given point.

Differential Revision: https://phabricator.services.mozilla.com/D27158

emilio committed:
style: Unify a bit Servo and Gecko code in properties.mako.rs.

Mostly removing #[cfg] statements, so npotb.

Commit:
style: [cssom][css-grid] 'grid-auto-flow: row dense' should serialize to 'dense' since 'row' is implied.

Differential Revision: https://phabricator.services.mozilla.com/D28058

emilio committed:
style: Don't keep two list of stylesheets in ServoStyleSet.

Just one set of stylesheets is enough. While at it, unify SheetType and Origin.

Differential Revision: https://phabricator.services.mozilla.com/D27564

emilio committed:
style: Don't allow to parse XUL tree pseudo-elements with a single colon.

Now that they're not exposed to the web we can remove this special case.

Differential Revision: https://phabricator.services.mozilla.com/D28071

emilio committed:
style: Drop unused user-agent cascade datas when not holding the cache lock.

We want to drop the cascade data memory as soon as possible, so bug 1544546
introduced an UpdateStylistIfNeeded call from ShellDetachedFromDocument.

Unfortunately, this call can reenter into the global user-agent cascade data if
some of the CSS values kept alive by the cascade data keep alive an SVG
document, see the stack on this bug for an example. Make sure to drop the
user-agent cascade datas when not holding the cache lock to avoid this
situation.

Before bug 1535788, we just destroyed the stylist, so we kept holding a
reference from the cache, and that reference will be dropped sometime later when
other document updated their user-agent stylesheets (if they happened not to
match the cache of course).

Seems to me this doesn't ended up happening in our automation, but it could
happen in the wild, in theory at least.

It's nice that Rust made this a safe deadlock caught by our tests rather than a
very subtle and infrequent memory corruption.

The relevant SVG documents are probably the <input type=number> rules:

https://searchfox.org/mozilla-central/rev/d80f0a570736dce76a2eb184fb65517462089e8a/layout/style/res/forms.css#1050

Differential Revision: https://phabricator.services.mozilla.com/D28299

emilio committed:
style: Allow CSS wide-keywords in custom property fallback.

Differential Revision: https://phabricator.services.mozilla.com/D28349

emilio committed:
style: Remove support for XBL resources.

So much unsound code going away :-)

Differential Revision: https://phabricator.services.mozilla.com/D28380

birtles committed:
style: Allow animating the 'all' property from Web Animations.

Differential Revision: https://phabricator.services.mozilla.com/D28763

emilio committed:
style: Check iterator length in SelectorIter::is_featureless_host_selector.

I'm going to unconditionally generate the PseudoElement combinator, and this
causes issues since we'll put the raw `::pseudo` selectors in the host bucket,
which is obviously wrong.

Differential Revision: https://phabricator.services.mozilla.com/D27528

emilio committed:
style: Fix :-moz-svg-use-shadow-tree-root pseudo-class.

Turns out removing the pseudo-class and such ends up not being quite as trivial
as I initially thought, or possible at all, since the fact that the <symbol> is
a <symbol> is observable via selectors, added a test for that.

Differential Revision: https://phabricator.services.mozilla.com/D29131

emilio committed:
style: Add a pref for a simpler -moz- gradient parsing.

This won't reintroduce any of the regressions that were triggered by our
previous attempts to turn off -moz prefixed gradients, and lets us massively
simplify the gradient code, if it sticks.

Differential Revision: https://phabricator.services.mozilla.com/D29346

emilio committed:
style: Use rust types for gradient stops.

This doesn't clean up all that much, yet, but it's a step in the right
direction.

Differential Revision: https://phabricator.services.mozilla.com/D29168

masayuki-nakano committed:
Move all remaining members of nsIPresShell to mozilla::PresShell

Additionally, this sorts out the order of member variables for minimizing the
instance size.

And also this changes `enum RenderFlags` to `enum class RenderingStateFlags`.

Differential Revision: https://phabricator.services.mozilla.com/D29312

emilio committed:
style: Use rust types for vertical-align.

The previous commit removed the dependence on the discriminant value, so we
don't need to keep discriminants different from text-align anymore.

Differential Revision: https://phabricator.services.mozilla.com/D29361

emilio committed:
style: Add parsing support for ::part().

Disabled for now of course. This should be pretty uncontroversial I'd think.

Differential Revision: https://phabricator.services.mozilla.com/D28060

emilio committed:
style: Implement selector-matching for ::part().

Also fairly straight-forward. This may get more complicated when we do part
forwarding, if any.

I've opened https://github.com/w3c/csswg-drafts/issues/3841 in what I think
would be a cleaner model for forwarding.

Differential Revision: https://phabricator.services.mozilla.com/D28063

emilio committed:
Remove a useless MallocSizeOf implementation.

Now that this is no longer used in AuthorStyles there's no point in doing this.

emilio committed:
style: Rustfmt recent changes.

jeremy-ir committed:
style: Convert NS_STYLE_BORDER to an enum class in nsStyleConsts.h.

Converting the NS_STYLE_BORDER definitions in to enumerated classes as
per bug 1277133.

The original constants broke the convention used by the rest of the
definitions as the CSS property being described is `border-collapse`,
so corrections were made with the migration to the enumerated class.

Differential Revision: https://phabricator.services.mozilla.com/D29951

Commit:
style: [css-grid-2] Remove single keyword 'subgrid' as a valid value for the 'grid' and 'grid-template' shorthands.

Differential Revision: https://phabricator.services.mozilla.com/D29974

emilio committed:
style: Fix servo build.

emilio committed:
layout: fix Servo build.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/element.rs
 * @kichjang: components/script/dom/element.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style, layout, and script code, but no tests are modified. Please consider adding a test!

emilio commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 04719e0 has been approved by `emilio`

<!-- @bors-servo r=emilio 04719e0acd148def1ae2025e5bab3251dd3ab750 -->

bors-servo commented:
:hourglass: Testing commit 04719e0acd148def1ae2025e5bab3251dd3ab750 with merge 756b2aac4c9ece50588ebbe795e6e45cf6ed9507...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12126)

CYBAI commented:
More tests passed! Looks like they're related to f81f848 which is the patch for `text-transform`!

```
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": null, 
    "test": "/css/css-variables/wide-keyword-fallback.html", 
    "line": 81448, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "e.style['text-transform'] = \"full-width\" should set the property value", 
    "test": "/css/css-text/parsing/text-transform-valid.html", 
    "line": 200453, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "e.style['text-transform'] = \"full-size-kana\" should set the property value", 
    "test": "/css/css-text/parsing/text-transform-valid.html", 
    "line": 200454, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "e.style['text-transform'] = \"capitalize full-width\" should set the property value", 
    "test": "/css/css-text/parsing/text-transform-valid.html", 
    "line": 200455, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "e.style['text-transform'] = \"uppercase full-size-kana\" should set the property value", 
    "test": "/css/css-text/parsing/text-transform-valid.html", 
    "line": 200456, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "e.style['text-transform'] = \"full-width full-size-kana\" should set the property value", 
    "test": "/css/css-text/parsing/text-transform-valid.html", 
    "line": 200457, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "e.style['text-transform'] = \"full-width lowercase\" should set the property value", 
    "test": "/css/css-text/parsing/text-transform-valid.html", 
    "line": 200458, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "e.style['text-transform'] = \"full-size-kana capitalize\" should set the property value", 
    "test": "/css/css-text/parsing/text-transform-valid.html", 
    "line": 200459, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "e.style['text-transform'] = \"full-size-kana full-width\" should set the property value", 
    "test": "/css/css-text/parsing/text-transform-valid.html", 
    "line": 200460, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "e.style['text-transform'] = \"capitalize full-width full-size-kana\" should set the property value", 
    "test": "/css/css-text/parsing/text-transform-valid.html", 
    "line": 200461, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "e.style['text-transform'] = \"full-width full-size-kana uppercase\" should set the property value", 
    "test": "/css/css-text/parsing/text-transform-valid.html", 
    "line": 200462, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "e.style['text-transform'] = \"full-size-kana lowercase full-width\" should set the property value", 
    "test": "/css/css-text/parsing/text-transform-valid.html", 
    "line": 200463, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "e.style['text-transform'] = \"lowercase full-size-kana full-width\" should set the property value", 
    "test": "/css/css-text/parsing/text-transform-valid.html", 
    "line": 200464, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "e.style['text-transform'] = \"full-width uppercase full-size-kana\" should set the property value", 
    "test": "/css/css-text/parsing/text-transform-valid.html", 
    "line": 200465, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "e.style['text-transform'] = \"full-size-kana full-width capitalize\" should set the property value", 
    "test": "/css/css-text/parsing/text-transform-valid.html", 
    "line": 200466, 
    "action": "test_result", 
    "expected": "FAIL"
}
```

emilio committed:
style: scroll-snap-coordinate shouldn't use NotInitial.

The initial value for this is indeed `none` (and thus empty). The Rust code was
confused.

This property is disabled by default these days, and I think the
get_initial_value() function, which is what could get confused, is not called
for this property, so I think this shouldn't be observable.

Differential Revision: https://phabricator.services.mozilla.com/D30124

emilio committed:
style: Update test expectations.

emilio commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 1bb15d8 has been approved by `emilio`

<!-- @bors-servo r=emilio 1bb15d881951425dd62f5d5b3677a18d1019167d -->

emilio commented:
/me *shrugs*
